### PR TITLE
Allow generating code in all Kotlin compile tasks

### PR DESCRIPTION
**Pull request by vRallev**
This change removes some workarounds for discovering contributed classes in the same module as components are merged. It also allows implementing other features that require more code generating in the future.

### CODE REVIEW

File: compiler/src/main/java/com/squareup/hephaestus/compiler/ClassScanner.kt [hunk 1]
```diff
@@ -42,50 +38,6 @@ internal class ClassScanner(
...
     }
   }
-
-  /**
-   * Returns a list of top level classes (meaning no inner classes) in this module. Note that the
-   * algorithm for this method takes all used packages from this module and lists all classes within
-   * this package. That means the returned list could include classes from module dependencies in the
-   * same package.
-   */
-  fun findTopLevelClassesInThisModule(
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/ClassScanner.kt:52]: This was strange from the beginning. I'm glad it's gone.

File: compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt [hunk 2]
```diff
@@ -21,17 +21,20 @@ class HephaestusComponentRegistrar : ComponentRegistrar {
...
     configuration: CompilerConfiguration
   ) {
-    val kotlinSourceRoots = configuration.kotlinSourceRoots.ifEmpty { return }
-    val scanner = ClassScanner(createPackageList(kotlinSourceRoots))
+    val scanner = ClassScanner()
     val sourceGenFolder = File(configuration.getNotNull(srcGenDirKey))
 
-    // The analysis phase is skipped when generating stubs for annotation processors. In unit tests
-    // we want to test this path, too.
-    if (!configuration.get(skipAnalysisKey, false)) {
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt:30]: Found the actual reason and this scenario doesn't need to be tested anymore.

File: compiler/src/main/java/com/squareup/hephaestus/compiler/InterfaceMerger.kt [hunk 3]
```diff
@@ -110,35 +105,4 @@ internal class InterfaceMerger(
...
   }
-
-  private fun Sequence<ClassDescriptor>.findInnerClasses(
-    thisDescriptor: ClassDescriptor,
-    scope: ClassDescriptor
-  ): Sequence<ClassDescriptor> {
-    return flatMap {
-      // It's necessary to exclude thisDescriptor and merged component interfaces, otherwise we
-      // could end up in nested loops while finding all classes and compiling the code. While
-      // compiling the code this extension would be called to add super classes.
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/InterfaceMerger.kt:121]: Very glad this isn't needed anymore.

File: compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt [hunk 4]
```diff
@@ -41,16 +44,46 @@ class HephaestusComponentRegistrar : ComponentRegistrar {
...
+    project: MockProject,
+    extension: AnalysisHandlerExtension
+  ) {
+    @Suppress("DEPRECATION")
+    val analysisHandlerExtensionPoint = Extensions.getArea(project)
+        .getExtensionPoint(AnalysisHandlerExtension.extensionPointName)
+
+    val registeredExtensions = AnalysisHandlerExtension.getInstances(project)
+    registeredExtensions.forEach {
+      // This doesn't work reliable, but that's the best we can do with public APIs. There's a bug
```
Comment (kirillzh) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt:57]: typo: reliable -> reliably.

Comment (vRallev) [reply compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt]: Fixed

File: compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt [hunk 5]
```diff
@@ -41,16 +44,46 @@ class HephaestusComponentRegistrar : ComponentRegistrar {
...
+  ) {
+    @Suppress("DEPRECATION")
+    val analysisHandlerExtensionPoint = Extensions.getArea(project)
+        .getExtensionPoint(AnalysisHandlerExtension.extensionPointName)
+
+    val registeredExtensions = AnalysisHandlerExtension.getInstances(project)
+    registeredExtensions.forEach {
+      // This doesn't work reliably, but that's the best we can do with public APIs. There's a bug
+      // for inner classes where they convert the given class to a String "a.b.C.Inner" and then
+      // try to remove "a.b.C$Inner". Good times! Workaround is below
```
Comment (FrostRocket) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/HephaestusComponentRegistrar.kt:59]: Nit: Forgot punctuation at the end.

File: compiler/src/main/java/com/squareup/hephaestus/compiler/Utils.kt [hunk 6]
```diff
@@ -45,7 +42,19 @@ internal fun ClassDescriptor.findAnnotation(
 ): AnnotationDescriptor? {
   // Must be JVM, we don't support anything else.
   if (!module.platform.has<JvmPlatform>()) return null
-  val annotationDescriptor = annotations.findAnnotation(annotationFqName)
+  val annotationDescriptor = try {
+    annotations.findAnnotation(annotationFqName)
+  } catch (e: IllegalStateException) {
+    // In some scenarios this exceptions is thrown. Throw a new exception with a better explanation.
```
Comment (FrostRocket) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/Utils.kt:48]: Nit: exception, not exceptions

File: compiler/src/main/java/com/squareup/hephaestus/compiler/Utils.kt [hunk 7]
```diff
@@ -45,7 +42,19 @@ internal fun ClassDescriptor.findAnnotation(
...
+    // Caused by: java.lang.IllegalStateException: Should not be called!
+    // at org.jetbrains.kotlin.types.ErrorUtils$1.getPackage(ErrorUtils.java:95)
+    throw HephaestusCompilationException(
```
Comment (FrostRocket) [inline compiler/src/main/java/com/squareup/hephaestus/compiler/Utils.kt:51]: Much better 👍

File: compiler/src/test/java/com/squareup/hephaestus/compiler/InterfaceMergerTest.kt [hunk 8]
```diff
@@ -198,8 +194,9 @@ class InterfaceMergerTest(
     """
     ) {
       assertThat(exitCode).isEqualTo(COMPILATION_ERROR)
-      // Position to the class.
-      assertThat(messages).contains("Source.kt: (13, 11)")
+      // Position to the class. Unfortunately, a different error is reported that the class is
+      // missing an @Module annotation.
+      assertThat(messages).contains("Source.kt: (7, 7)")
```
Comment (FrostRocket) [inline compiler/src/test/java/com/squareup/hephaestus/compiler/InterfaceMergerTest.kt:199]: Wouldn't it be more useful to search for something more meaningful than the source lines? Might be annoying if you're hardcoding the line numbers and it changes on you.

Comment (vRallev) [reply compiler/src/test/java/com/squareup/hephaestus/compiler/InterfaceMergerTest.kt:199]: When I started writing these tests, the error messages slightly changed here and there. To me it was more important that the error is reported in the correct row and column and points to the offending piece of code. But at the same time I agree with you. That's a good opportunity to contribute :-)

**Comment by vRallev** [comment]
Addressed feeback.

**Pull request merged**